File: Graph/Longest Path/Unweighted graph/Java/Graph.java

// Steps:-
//1) Perform a dFS run from each vertex and also keep storing the maximum recorded path from a vertex in a 
//   memo/cache table so that it can be used in subsequent DFS runs
//2) so dfs(i) would be -> return (memo[i] = 1+ Max (dfs (i1), dfs(i2), dfs(i3).....)
//3) for all directed edges from i to i1 , i2,..i3...and storing the max recorded path into the memo table

import java.util.*;

class Solution {
    private int maxLen = 0;
    private int[] memo;
    public int solve(int[][] graph) {
        if (graph == null || graph.length == 0)
            return 0;
        memo = new int[graph.length + 1];
        Arrays.fill(memo, -1);

        for (int i = 0; i < graph.length; i++) maxLen = Math.max(maxLen, dfs(graph, i));
        return maxLen - 1;
    }

    private int dfs(int[][] graph, int vertex) {
        if (memo[vertex] != -1)
            return memo[vertex];
        int maxPath = 0;
        for (int adjacentVertex : graph[vertex]) {
            maxPath = Math.max(maxPath, dfs(graph, adjacentVertex));
        }
        return memo[vertex] = 1 + maxPath;
    }
}

File: Graph/Longest Path/Unweighted graph/cpp/Graph.cpp
// Problem :- To find the longest path in the directed unweighted (acyclic) graph 

 
// Steps:- 
// 1) it is given that graph is acyclic and unweighted directed.
// 2) We only need to find maximum depth a recursion (dfs) can go, that will give us max length.
// 3) Add memoization to avoid recomputing.

int dfs(vector<vector<int>>& graph, int s, vector<int>& dis) {
    if (dis[s] != -1) return dis[s];

    if (graph[s].size() == 0) return dis[s] = 0;

    int ans = 0;
    for (auto v : graph[s]) {
        ans = max(ans, dfs(graph, v, dis));
    }

    return dis[s] = ans + 1;
}
int solve(vector<vector<int>>& graph) {
    int n = graph.size();

    int len = 0;
    vector<int> dis(n, -1);
    for (int i = 0; i < n; i++) {
        dis[i] = dfs(graph, i, dis), len = max(len, dis[i]);
    }
    // for(int i=0;i<n;i++)cout<<dis[i]<<" ";
    return len;
}
